### ISSUE TITLE
Documentation improvements needed. Instructions unclear.

### ISSUE THREAD

**Issue by svenevs** [niosus/EasyClangComplete#274]
Hello,

I'm very new to this plugin, as well as explicit sublime project files.  I'm not quite sure if this is an "issue" for you, or if what I am looking for is not (currently) possible.  I've been browsing around the examples / other issues here, and it seems like the general approach here is to hard-code include paths (e.g. for mingw, or [this line](https://github.com/niosus/EasyClangComplete/blob/6ca6ccb72edb35ebcd418978c60f002822b16062/EasyClangComplete.sublime-settings#L21)).

To be clear, I'm still struggling to get a generic project working with `#include <iostream>`, even with hard-coded paths.  So the actual question here is really just "does this tool require hard-coded paths e.g. `/usr/include` to work?"  I'm trying to setup a project that has many users with widely varied ecosystems, in particular, many things that show up in `/usr/include` etc are incorrect.  We use various modules that rpath / prepend to the relevant path utilities.  I suspect that, even though ECC can find my `clang` executable, since it is isolated in the trenches of `/opt`, some of these defaults may actually be causing the conflict?  It's a full source build of `clang 4.0.0` if that matters.

As a corollary, based on some testing, EasyClangComplete will still work / not care if we hard-code paths for system A that do not exist on system B?

I guess my confusion here is really just what is the difference between what ECC is grabbing from `CMakeLists.txt`, vs this `.clang_complete` file, and the submlime settings?  It seems that the order is

1. User / package settings.
2. Override any of those with specific project settings.
3. ? `.clang_complete` definitions loaded.
4. `CMakeLists.txt` parsed?  Or does the eventual include path generated by `cmake` not actually have any effect?

Sorry if these answers are available already, I've been tinkering with this on and off for a couple of days and feel like I'm really missing the bigger picture here.  Any pointers are greatly appreciated, even if its just "keep searching, you'll find it" 😱

**Comment by niosus** [comment]
Ok, so the fact that you have to ask this question shows that I did not do a good job at providing enough information about the configurations. Let me try to summarize things here really quick.

## How flags are read
There are two options on how to generate/load needed compilation flags: 
- from external sources
- from plugin setting `common_flags`
Let's look into these:
### External sources
- generate them from cmake.
- load them from a compilation database (generated by, say, [Bear](https://github.com/rizsotto/Bear)).
- load them from `.clang_complete` file.

It is important to understand here how the plugin picks the source from which to load the flags. For example, the setting `flags_sources` of the plugin is the following:
```json
    "flags_sources": [
        {"file": "CMakeLists.txt"},
        {"file": "compile_commands.json"},
        {"file": ".clang_complete"},
    ],
```
This will force the plugin to first search for a valid `CMakeLists.txt` with a line `project(***)` in it. 

If it finds such a file, the search is stopped and the flags are loaded by running cmake over it and reading the produced compilation database.

If the file is not found, the plugin switches to the next specified source and tries to search for a `compile_commands.json` file in project folder (or user specified folder). Again, if the file is found, the flags get loaded from it and the search is stopped.

Finally, if the compilation database was not found, the plugin will try to search for `.clang_complete` file and everything else is just like before.

The order of this search is defined solely by `flags_sources` setting (also can be `ecc_flags_sources` in project-specific settings)

This is all to know about flag generation from external sources.  

### Plugin settings
- define flags in `common_flags` setting in the plugin
- or redefine them in `ecc_common_flags` setting in the project-specific settings. The **important** part about this option is that these override the `common_flags` from generic settings.
### How these sources merge
Whatever we've picked from the above the following holds:

**Flags from plugin settings get merged with the flags from external sources**

## Out of the box settings
The plugin should work out of the box on any cmake project on any system.

## Why hardcoding any flags at all?
Clang cannot find all the stl headers on all systems on its own unfortunately. Therefore some hardcoded paths are sometimes needed. There should generally be no problem if there is a non-existing include folder provided as a flag as clang just ignores it.

Unfortunately, I cannot tell more about how the thing works on Windows. I mostly work on Linux and switch to Windows only to test this plugin to be honest, so I can't be of too much help there, especially with MinGW and co.

Ok, so, did I clear it at least a little bit for you? If yes, would you have suggestions on how this should be presented (or what should be added/removed to/from the plugin) to aid understanding? 

I want it to be an intuitive tool, that is why I started it in the first place.

**Comment by svenevs** [comment]
Ahhhhhhhhh much clearer indeed!

> Flags from plugin settings get merged with the flags from external sources

In addition to the `flags_sources` defining the order, that is definitely something that makes it much clearer.  No replacements (good god, I can't even imagine trying to implement that).  Order hierarchy is clear.  E.g. if I changed my project to be

```json
    "flags_sources": [
        {"file": ".clang_complete"},
        {"file": "CMakeLists.txt"},
    ],
```

Then it would

1. First, look in `.clang_complete` now.
2. Second, do the magic with `cmake`.
3. Even if `compile_commands.json` is present, it will be ignored because we overrode the package defaults.

> hardcoding any flags at all

First, a clarifying question: assuming the user has a valid `CMakeLists.txt` and their environment is setup such that any `find_package(FOO)` is actually found, does ECC complete through these?

So I noticed [`libclang_path`](https://github.com/niosus/EasyClangComplete/blob/6ca6ccb72edb35ebcd418978c60f002822b16062/EasyClangComplete.sublime-settings#L103), and think resolving my "situation" could be remedied by allowing a more generic `llvm_root` option or something.  I'm not 100% on this, but I think they always have the same structure.  Generally speaking, though, I doubt many users are working with a full-blown `llvm` toolkit.  I was able to get my project to start finding things by adding the following:

```json
"ecc_common_flags": [
    "-I/opt/crazy/long/path/llvm-4.0.0/include/c++/v1/",
    "-I/opt/crazy/long/path/llvm-4.0.0/include/clang",
]
```

I ask about `find_package(FOO)` because I think the toolchain I'm using is really the only ECC doesn't explicitly support -- I was playing around on OSX and also whipped up a VM with vanilla `dnf install clang` and everything totally works.  If you have a user that has a full-blown `llvm` stack compiled, on the other hand, it is "guaranteed" that this will be installed in a self-contained directory.  Two things I observe

```console
$ echo $CMAKE_PREFIX_PATH | sed 's/:/\n/g' | grep -i llvm
/opt/crazy/long/path/llvm-4.0.0/

$ which clang
/opt/crazy/long/path/llvm-4.0.0/bin/clang
```

So while parsing `CMAKE_PREFIX_PATH` could potentially solve this, it sounds like a nightmare.  If you add a config variable `llvm_root`, you can then default to looking in

- `$llvm_root/bin/$clang_binary`
- `$llvm_root/include/c++/v1` -- not sure if the `v1` is always there though
- `$llvm_root/include/` -- are these files even necessary?

If helpful, I made have the [`tree` output here](https://gist.github.com/svenevs/71ce2dca23a98338c0a4e2a679cb2feb).  I'm not entirely sure what you can / cannot rely on from that, but it may be helpful for identifying what would even need to be specified.

Thank you for such a thoughtful response!  I think that just adding the above explanation of what comes from where in the main readme would help newbies like me understand what could be wrong / how to troubleshoot 😄

**Comment by niosus** [comment]
I definitely will refubrish the readme as answering your question forced me to structure it better in my brain too. :smile: Writing documentation is not the most inspiring thing ever :)

I will need some more time to answer everything else from your response. But will do during the week, I guess.

**Comment by niosus** [comment]
Closed with #514
